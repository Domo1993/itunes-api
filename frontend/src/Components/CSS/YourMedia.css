#mediaHead {
    background: lightgray;
    padding-bottom: 50px;
}

#play {
    float: left;
    max-width: 11%;
    height: auto;
    margin-left: 20%;
    margin-top: 2%;
}

.playImage:hover {
    animation: rotation 2s infinite linear;
  }
  
  
  @keyframes rotation {
    from {
      transform: rotate(0deg);
    }
    to {
      transform: rotate(359deg);
    }
  }

#media {
    font-family: 'Nova Flat';
    font-size: 5vw;
    color: rgb(245, 24, 24);
    padding-top: 20px;
    text-shadow: 3px 3px 3px black;
}

#subtitle {
    
}

.links {
    float: right;
    text-decoration: none;
    background-color: rgba(24, 23, 109, 0.637);
    color: white;
    text-shadow: 1px 1px 1px black;
    padding: 10px;

}

.links:hover {
    text-decoration: none;
}

form {
    margin-left: auto;
    margin-right: auto;
    text-align: center;
    background-color: grey;
    padding: 20px;
    max-width: 50%;
    border-radius: 3em;
    margin-bottom: 2%;
}

#help {
    font-weight: lighter;
    color: white;
}

select,option {
    height: 31px;
}

.fa, .fa-star {
    color:gold;
    font-size: 20px;
}

.tableCol {
    text-align: center;
    font-family: 'Dekko';
}


tbody:nth-child(odd) {
    background-color: rgba(252, 185, 185, 0.651);
}

tbody:nth-child(even) {
    background-color: rgba(69, 89, 126, 0.493);
}

* {
    outline: none;
}

.dataTableDiv {
    text-align: center;
    width: 100%;
    margin-left: auto;
    margin-right: auto;
}

table,tr,td {
    padding: 10px;
    margin-left: auto;
    margin-right: auto;
    
}

