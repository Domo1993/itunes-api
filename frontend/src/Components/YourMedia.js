import React, { Component } from 'react';
import { NavLink } from 'react-router-dom';
// cleaning the json files text's readability
import parse from 'html-react-parser';
// Importing components
import DescriptionModal from './DescriptionModal.js';
import VideoModal from './VideoModal.js';

import './CSS/YourMedia.css';
import play from './Images/redplay.png';

class YourMedia extends Component {
    constructor(props) {
        super(props);
        // Setting intitial states
        this.state = {
            searchResults: [],
            search: "",
            type: "",
            favouritesList: []
        };
    }

    // The main search function which depends on parameters from the user's input
    mySearch = async () => {
        const search = this.state.search;
        const type = this.state.type;
        
        const api_call = await fetch(`/api?term=${search}&entity=${type}`);
        const response = await api_call.json();
        
        this.setState({
            searchResults: response
        })
    };

    // This function is used to send the favourites information to the backend and return the info into an array
    // It also checks if a favourite has been added already and will indicate by an alert.
    myPost = async (favId,collectName, image, artist, title, preview, wrapper, kind, descrip) => {
        const favouritesCall = await fetch(`/favourites`);
        const favJson = await favouritesCall.json();

        let added = false

        if(favJson.length === 0){
            fetch(`/favourites?id=${favId}&collectionName=${collectName}&image=${image}&artist=${artist}&title=${title}&previewUrl=${preview}&wrapperType=${wrapper}&kind=${kind}&description=${descrip}`, {
                method: "POST", headers: { "Content-Type": "application/json" } })
            .then(() => {
                fetch('/favourites', { method: "GET", headers: { "Content-Type": "application/json" } })
                    .then(res => res.json())
                    .then(favouritesList => this.setState({ favouritesList }));
                })
            alert("Added to Your Favourites!")
        } else {
            for(let i=0;i<favJson.length;i++){

                if(favJson[i].id == favId || favJson[i].previewUrl === preview){
                    console.log(i)
                    alert("Already added to favourites...")
                    added = true
                     }} if (added === false) {
                        fetch(`/favourites?id=${favId}&collectionName=${collectName}&image=${image}&artist=${artist}&title=${title}&previewUrl=${preview}&wrapperType=${wrapper}&kind=${kind}&description=${descrip}`, {
                            method: "POST", headers: { "Content-Type": "application/json" } })
                        .then(() => {
                            fetch('/favourites', { method: "GET", headers: { "Content-Type": "application/json" } })
                                .then(res => res.json())
                                .then(favouritesList => this.setState({ favouritesList }));
                            })
                        alert("Added to Your Favourites!")
                    }                   
        }   
    }
    // This function will return either an audio tag or video tag(within a modal) depending on the user's search
    mediaCheck = (media, other, source) => {
                if(media === "song" || media === "podcast" || other === "audiobook"){
                    return <audio src={source} controls/>
                } else if(media === "music-video" || media === "feature-movie" || media === "tv-episode"){
                    return <VideoModal video={<video width="320" height="240" src={source} controls> Sorry, your browser doesn't support HTML5</video>} />
                }
            }

    // This function returns the description (within a modal) of an audio book if the user searches audiobook
    audioBookDescrip = (wrapper,descrip) => {
        if(wrapper === "audiobook"){
            return <DescriptionModal description={parse(`${descrip}`)} />
        }        
    }

    // This function returns the title of the user's searched item/s
    audioBookTitleCheck = (wrapper,collectName,songName) => {
        if(wrapper === "audiobook"){
            return <h4>{collectName}</h4>
        } else {
            return <h4>{songName}</h4>
        }
    }

    render() {
        return (
            <div className="mainDiv">
                
                <header id="mediaHead">
                    <img className="playImage" id="play" src={play} alt="play button"/>
                    <h1 id="media" >YOUR MEDIA</h1>
                    <p id="subtitle">Songs, Music Videos, Movies, and more... </p>
                    <NavLink className="links" to="/Favourites_list">Your Favourites</NavLink>
                </header>

                <br/>
                {/* The main search and dropdown menu for the user to interact with */}
                <form className="form">
                    <p id="help">Use the dropdown menu to select a media type</p>
                    <input onChange={e => this.setState({search: e.target.value})} type="text" name="search" placeholder="Search"/>

                    <select name="media_type" onChange={e => this.setState({type: e.target.value})}>
                        <option value="song">Song </option>
                        <option value="musicVideo">Music Video </option>
                        <option value="movie">Movie </option>
                        <option value="tvEpisode">Tv Episodes </option>
                        <option value="audiobook">Audio Book </option>
                    </select>

                    <button onClick={(e) => {
                        this.mySearch() 
                        e.preventDefault()
                    }} type="submit" value="Submit" id="btn1">Search</button>
                </form>

            <div className="dataTableDiv">
                <table className="dataTable" id="mainTable">
                    {/*Mapping all the data into a table*/}
                    {this.state.searchResults.map((obj, key) => {
                        return (
                            <tbody key={key++} className="tableRow">
                                <tr>
                                  <td className="tableCol"><img className="artworks" src={obj.artworkUrl100} alt="artwork"/></td>
                                  <td className="tableCol">{this.audioBookTitleCheck(obj.wrapperType,obj.collectionName,obj.trackName)}</td>
                                  <td className="tableCol">{obj.artistName}</td>                                  
                                  <td className="tableCol">{this.mediaCheck(obj.kind, obj.wrapperType, obj.previewUrl)}
                                    {this.audioBookDescrip(obj.wrapperType,obj.description)}
                                  </td>
                                  <td className="tableCol">
                                    <button onClick={() => this.myPost(obj.trackId,obj.collectionName, obj.artworkUrl100, obj.artistName, obj.trackName, obj.previewUrl, obj.wrapperType, obj.kind, obj.description)}>
                                    <i className="fa fa-star"></i>
                                    </button>
                                  </td>
                                </tr>
                            </tbody>
                        );
                    })}
                </table>
            </div> {/*Table div end*/}


            </div>
        )
    }
}

export default YourMedia;
