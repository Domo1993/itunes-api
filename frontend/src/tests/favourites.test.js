import React from 'react';
import YourMedia from '../Components/YourMedia.js';
import renderer from 'react-test-renderer';

test('renders correctly', () => {
const tree = renderer
.create(<YourMedia/>)
.toJSON();
expect(tree).toMatchSnapshot();
});