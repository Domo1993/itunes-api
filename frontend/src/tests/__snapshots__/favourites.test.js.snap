// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="mainDiv"
>
  <header
    id="mediaHead"
  >
    <img
      alt="play button"
      className="playImage"
      id="play"
      src="redplay.png"
    />
    <h1
      id="media"
    >
      YOUR MEDIA
    </h1>
    <p
      id="subtitle"
    >
      Songs, Music Videos, Movies, and more... 
    </p>
    <a
      className="links"
      href="/Favourites_list"
    >
      Your Favourites
    </a>
  </header>
  <br />
  <form
    className="form"
  >
    <p
      id="help"
    >
      Use the dropdown menu to select a media type
    </p>
    <input
      name="search"
      onChange={[Function]}
      placeholder="Search"
      type="text"
    />
    <select
      name="media_type"
      onChange={[Function]}
    >
      <option
        value="song"
      >
        Song 
      </option>
      <option
        value="musicVideo"
      >
        Music Video 
      </option>
      <option
        value="movie"
      >
        Movie 
      </option>
      <option
        value="tvEpisode"
      >
        Tv Episodes 
      </option>
      <option
        value="audiobook"
      >
        Audio Book 
      </option>
    </select>
    <button
      id="btn1"
      onClick={[Function]}
      type="submit"
      value="Submit"
    >
      Search
    </button>
  </form>
  <div
    className="dataTableDiv"
  >
    <table
      className="dataTable"
      id="mainTable"
    />
  </div>
   
</div>
`;
